From 62be23ef4da65bffd6e75b9d12bc6db1ad4a3d8c Mon Sep 17 00:00:00 2001
From: Robby Cai <robby.cai@nxp.com>
Date: Thu, 20 Apr 2017 17:17:42 +0800
Subject: [PATCH 2375/5242] MLK-14720 epdc: correct WFE setting when bypass
 legacy process

commit  76e511065867777dcccff53f03677f5a6ff4e531 from
https://source.codeaurora.org/external/imx/linux-imx.git

set WFE (WFE_A on imx7d, and WFE_B on imx6ull/imx6sll) input address to
framebuffer start address, and set left/top coordinate since the framebuffer is
the original source of WFE (i.e., not from PXP output) when bypass legacy mode.
The patch also limits the condition to bypass legacy mode when not use
EPDC_FLAG_USE_ALT_BUFFER.

Signed-off-by: Robby Cai <robby.cai@nxp.com>
(cherry picked from commit 7f19940705902623166777c675f5e10c9e7fc477)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/video/fbdev/mxc/mxc_epdc_v2_fb.c |   14 +++++++++++++-
 1 file changed, 13 insertions(+), 1 deletion(-)

diff --git a/drivers/video/fbdev/mxc/mxc_epdc_v2_fb.c b/drivers/video/fbdev/mxc/mxc_epdc_v2_fb.c
index 14bcb76..b908ad7 100644
--- a/drivers/video/fbdev/mxc/mxc_epdc_v2_fb.c
+++ b/drivers/video/fbdev/mxc/mxc_epdc_v2_fb.c
@@ -2,6 +2,8 @@
  * Copyright (C) 2014-2016 Freescale Semiconductor, Inc.
  * Copyright 2017 NXP
  *
+ * Copyright 2017 NXP
+ *
  * This program is free software; you can redistribute it and/or modify
  * it under the terms of the GNU General Public License as published by
  * the Free Software Foundation; either version 2 of the License, or
@@ -6408,9 +6410,19 @@ static int pxp_wfe_a_process(struct mxc_epdc_fb_data *fb_data,
 		if ((fb_data->epdc_fb_var.rotate == FB_ROTATE_UR) &&
 			(fb_data->epdc_fb_var.grayscale == GRAYSCALE_8BIT) &&
 			!is_transform && (proc_data->dither_mode == 0) &&
+			!(upd_data_list->update_desc->upd_data.flags &
+			EPDC_FLAG_USE_ALT_BUFFER) &&
 			!fb_data->restrict_width) {
-				pxp_conf->wfe_a_fetch_param[0].paddr =
+			sg_dma_address(&sg[0]) = fb_data->info.fix.smem_start;
+			sg_set_page(&sg[0],
+				virt_to_page(fb_data->info.screen_base),
+				fb_data->info.fix.smem_len,
+				offset_in_page(fb_data->info.screen_base));
+			pxp_conf->wfe_a_fetch_param[0].paddr =
 					sg_dma_address(&sg[0]);
+
+			pxp_conf->wfe_a_fetch_param[0].left = update_region->left;
+			pxp_conf->wfe_a_fetch_param[0].top = update_region->top;
 		} else
 			pxp_conf->wfe_a_fetch_param[0].paddr =
 				upd_data_list->phys_addr + upd_data_list->update_desc->epdc_offs;
-- 
1.7.9.5

